res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,741 @@
+    engine.setValue(PioneerDDJ400.selectedFxGroup, "next_effect", value);
+};
+
+PioneerDDJ400.beatFxLeftPressed = function(_channel, _control, value) {
+    if (value === 0) { return; }
+
+    PioneerDDJ400.selectedFxSlot = 1;
+};
+
+PioneerDDJ400.beatFxRightPressed = function(_channel, _control, value) {
+    if (value === 0) { return; }
+
+    PioneerDDJ400.selectedFxSlot = 2;
+};
+
+PioneerDDJ400.beatFxSelectPressed = function(_channel, _control, value) {
+    if (value === 0) { return; }
+
+    PioneerDDJ400.selectedFxSlot = 3;
+};
+
+PioneerDDJ400.beatFxOnOffPressed = function(_channel, _control, value) {
+    if (value === 0) { return; }
+
+    // toggle the currently focused effect slot in Effect Unit 1 (if any)
+    var selectedSlot = PioneerDDJ400.selectedFxSlot;
+    if (selectedSlot <= 0 || selectedSlot > PioneerDDJ400.numFxSlots) {
+        return;
+    }
+    var isEnabled = !engine.getValue(PioneerDDJ400.selectedFxGroup, "enabled");
+    engine.setValue(PioneerDDJ400.selectedFxGroup, "enabled", isEnabled);
+    PioneerDDJ400.toggleLight(PioneerDDJ400.lights.beatFx, isEnabled);
+};
+
+PioneerDDJ400.beatFxOnOffShiftPressed = function(_channel, _control, value) {
+    if (value === 0) { return; }
+
+    // turn off all three effect slots in Effect Unit 1 and reset wet/dry mix
+    for (var i = 1; i <= PioneerDDJ400.numFxSlots; i += 1) {
+        engine.setValue("[EffectRack1_EffectUnit1_Effect" + i + "]", "enabled", 0);
+    }
+    script.triggerControl("[EffectRack1_EffectUnit1]", "mix", 0);
+    PioneerDDJ400.toggleLight(PioneerDDJ400.lights.beatFx, false);
+};
+
+PioneerDDJ400.beatFxChannel = function(_channel, control, value, _status, group) {
+    if (value === 0x00) { return; }
+
+    var enableChannel1 = control === 0x10 ? 1 : 0,
+        enableChannel2 = control === 0x11 ? 1 : 0,
+        enableMaster = control === 0x14 ? 1 : 0;
+
+    engine.setValue(group, "group_[Channel1]_enable", enableChannel1);
+    engine.setValue(group, "group_[Channel2]_enable", enableChannel2);
+    engine.setValue(group, "group_[Master]_enable", enableMaster);
+};
+
+//
+// Loop IN/OUT ADJUST
+//
+
+PioneerDDJ400.toggleLoopAdjustIn = function(channel, _control, value, _status, group) {
+    if (value === 0 || engine.getValue(group, "loop_enabled" === 0)) {
+        return;
+    }
+    PioneerDDJ400.loopAdjustIn[channel] = !PioneerDDJ400.loopAdjustIn[channel];
+    PioneerDDJ400.loopAdjustOut[channel] = false;
+};
+
+PioneerDDJ400.toggleLoopAdjustOut = function(channel, _control, value, _status, group) {
+    if (value === 0 || engine.getValue(group, "loop_enabled" === 0)) {
+        return;
+    }
+    PioneerDDJ400.loopAdjustOut[channel] = !PioneerDDJ400.loopAdjustOut[channel];
+    PioneerDDJ400.loopAdjustIn[channel] = false;
+};
+
+PioneerDDJ400.setReloopLight = function(status, value) {
+    midi.sendShortMsg(status, 0x4D, value);
+    midi.sendShortMsg(status, 0x50, value);
```

Comment:
Why are two MIDI messages required for one LED?

Reply:
This was one of the issues you helped me detect in zulip... the controller using different control bytes when shift is pressed.

Reply:
That is not obvious, so please add a comment.